assets/src/models/routeTab.ts

```diff
@@ -40,11 +46,108 @@ export const parseRouteTabData = (
 ): RouteTab[] => {
   return routeTabsData.map((routeTabData) => ({
     uuid: routeTabData.uuid,
-    ordering: routeTabData.ordering,
-    presetName: routeTabData.preset_name,
+    ordering: nullToUndefined(routeTabData.ordering),
+    presetName: nullToUndefined(routeTabData.preset_name),
     isCurrentTab: routeTabData.is_current_tab || false,
     selectedRouteIds: routeTabData.selected_route_ids,
     ladderDirections: routeTabData.ladder_directions,
     ladderCrowdingToggles: routeTabData.ladder_crowding_toggles,
   }))
 }
+
+export const isPreset = (routeTab: RouteTab): boolean =>
+  routeTab.presetName !== undefined
+export const isOpenTab = (routeTab: RouteTab): boolean =>
+  routeTab.ordering !== undefined
+
+export const instantiatePresetByUUID = (
+  routeTabs: RouteTab[],
+  uuid: string
+): RouteTab[] => {
+  const preset = routeTabs.find((routeTab) => routeTab.uuid === uuid)
+
+  if (preset === undefined) {
+    throw new Error(`No preset found for UUID ${uuid}`)
+  } else if (preset.ordering !== undefined) {
+    return routeTabs.map((routeTab) => {
+      if (routeTab.uuid === uuid) {
+        return { ...routeTab, isCurrentTab: true }
+      } else {
+        return { ...routeTab, isCurrentTab: false }
+      }
+    })
+  } else if (currentRouteTab(routeTabs).selectedRouteIds.length === 0) {
+    return routeTabs
+      .filter((routeTab) => routeTab.uuid !== uuid)
+      .map((routeTab) => {
+        if (routeTab.isCurrentTab) {
+          return { ...preset, ordering: routeTab.ordering, isCurrentTab: true }
+        } else {
+          return routeTab
+        }
+      })
+  } else {
+    return routeTabs.map((routeTab) => {
+      if (routeTab.uuid === uuid) {
+        return {
+          ...routeTab,
+          isCurrentTab: true,
+          ordering: highestExistingOrdering(routeTabs) + 1,
+        }
+      } else {
+        return { ...routeTab, isCurrentTab: false }
+      }
+    })
+  }
+}
+
+export const closeTabByUUID = (
+  routeTabs: RouteTab[],
+  uuid: string
+): RouteTab[] => {
+  const tabToClose = routeTabs.find((routeTab) => routeTab.uuid === uuid)
+
+  if (tabToClose === undefined) {
+    throw new Error(`No preset found for UUID ${uuid}`)
+  }
+
+  const newRouteTabs =
+    tabToClose.presetName === undefined
+      ? routeTabs.filter((routeTab) => routeTab.uuid !== uuid)
+      : routeTabs.map((routeTab) => {
+          if (routeTab.uuid === uuid) {
+            return { ...routeTab, ordering: undefined, isCurrentTab: false }
+          } else {
+            return routeTab
+          }
+        })
+
+  if (tabToClose.isCurrentTab) {
+    const nextTabToRight = newRouteTabs
+      .filter(
+        (routeTab) =>
+          routeTab.ordering && routeTab.ordering > (tabToClose.ordering || 0)
+      )
+      .sort((a, b) => a.ordering || 0 - (b.ordering || 0))[0]
+
+    if (nextTabToRight) {
```

Comment:
**thought:** I wonder if you could use `highestExistingOrdering` combined with the order of the `tabToClose` to check whether there is a tab to the right, without having to do the filtering and sorting first.

Reply:
Hmm, I'll have to consider that. Overall this bit of logic here is definitely the hairiest part of the PR so I'm definitely open to ideas for how to make it less confusing.

Reply:
I pre-prefix this by saying that this doesn't feel important for this PR.

Then I'll prefix it by saying that the following is an initial thought that I haven't investigated at all. It's quite likely it won't make sense in practice.

That said, part of what that initial thought came from was the notion that maybe you could bury all this logic in a helper function, called `nextTab` or `fallbackTab` or something that made more sense. Sort of a refactor to both hide but also "describe with a function name" kind of approach. Then maybe `highestExistingOrdering` could be checked and if there was a higher order number than the current `tabToClose`, that means there's a tab to the right (I think?) and you could just call this existing "next tab to right" logic, otherwise, you could call the "next tab to left" logic. It doesn't change a lot, but maybe it's a hair clearer? And with a function, you could test the logic in isolation.

Or just decide none of this is worth the effort and just ignore it.

Reply:
Over all I like that idea (and you have a good point about function names as documentation), but I'm not sure it's really worth the effort for this PR. Thanks for the idea, though!